File: neetcode/49groupAnagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 *
 * Input: strs = ["eat","tea","tan","ate","nat","bat"]
 * Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
 */
var groupAnagrams = function (strs) {
  const map = new Map();

  for (let i = 0; i < strs.length; i++) {
    const sortedWord = strs[i].split('').sort().join('');

    if (map.has(sortedWord)) {
      map.get(sortedWord).push(strs[i]);
    } else {
      map.set(sortedWord, [strs[i]]);
    }
  }

  return Array.from(map.values()).sort((a, b) => a.length - b.length);
};

const strs = ['eat', 'tea', 'tan', 'ate', 'nat', 'bat'];
console.log(groupAnagrams(strs));

File: neetcode/49groupAnagrams.py
from collections import defaultdict

def groupAnagrams(strs):
    anagrams = defaultdict(list)
    
    for word in strs:
        sorted_word = ''.join(sorted(word))
        anagrams[sorted_word].append(word)
    
    return sorted(anagrams.values(), key=len)

# Example usage
strs = ['eat', 'tea', 'tan', 'ate', 'nat', 'bat']
print(groupAnagrams(strs))